src/Agent.Plugins/PipelineCache/PipelineCacheServer.cs

```diff
@@ -42,32 +52,150 @@ public class PipelineCacheServer
-                    actionAsync: async () =>
+                if (!isTar) // Uploading as files.
+                {
+                    //Upload the pipeline artifact.
+                    PipelineCacheActionRecord uploadRecord = clientTelemetry.CreateRecord<PipelineCacheActionRecord>((level, uri, type) =>
+                        new PipelineCacheActionRecord(level, uri, type, nameof(dedupManifestClient.PublishAsync), context));
+                    PublishResult result = await clientTelemetry.MeasureActionAsync(
+                        record: uploadRecord,
+                        actionAsync: async () =>
+                        {
+                            return await dedupManifestClient.PublishAsync(path, cancellationToken);
+                        });
+
+                    CreatePipelineCacheArtifactOptions options = new CreatePipelineCacheArtifactOptions
                     {
-                        return await dedupManifestClient.PublishAsync(path, cancellationToken);
-                    });
+                        Fingerprint = fingerprint,
+                        RootId = result.RootId,
+                        ManifestId = result.ManifestId,
+                        ProofNodes = result.ProofNodes.ToArray(),
+                    };
+
+                    // Cache the artifact
+                    PipelineCacheActionRecord cacheRecord = clientTelemetry.CreateRecord<PipelineCacheActionRecord>((level, uri, type) =>
+                        new PipelineCacheActionRecord(level, uri, type, PipelineArtifactConstants.SaveCache, context));
+                    CreateStatus status = await pipelineCacheClient.CreatePipelineCacheArtifactAsync(options, cancellationToken, cacheRecord);
 
-                CreatePipelineCacheArtifactOptions options = new CreatePipelineCacheArtifactOptions
+                    // Send results to CustomerIntelligence
+                    context.PublishTelemetry(area: PipelineArtifactConstants.AzurePipelinesAgent, feature: PipelineArtifactConstants.PipelineCache, record: uploadRecord);
+                    context.PublishTelemetry(area: PipelineArtifactConstants.AzurePipelinesAgent, feature: PipelineArtifactConstants.PipelineCache, record: cacheRecord);
+                }
+                else
                 {
-                    Fingerprint = fingerprint,
-                    RootId = result.RootId,
-                    ManifestId = result.ManifestId,
-                    ProofNodes = result.ProofNodes.ToArray(),
-                };
+                    var archiveFile = Path.Combine(Path.GetTempPath(), Guid.NewGuid().ToString() + "archive.tar");
+                    var processTcs = new TaskCompletionSource<int>();
+                    using (var cancelSource = new CancellationTokenSource())
+                    using (var linkedSource = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, cancelSource.Token))
+                    using (var process = new Process())
```

Comment:
We need to create a TAR class that is specific to pipeline caching so we can modularize the system. We should have a method that does extraction and archiving.